src/Fantomas.Client/LSPFantomasService.fs

```diff
@@ -150,24 +174,33 @@ let private daemonNotFoundResponse filePath : Task<FantomasResponse> =
       Content = Some(sprintf "No fantomas tool was found for \"%s\"" filePath) }
     |> Task.FromResult
 
+let private cancellationWasRequestedResponse filePath : Task<FantomasResponse> =
+    { Code = int FantomasResponseCode.CancellationWasRequested
+      FilePath = filePath
+      Content = Some "FantomasService is being or has been disposed." }
+    |> Task.FromResult
+
 let mapResultToResponse (filePath: string) (result: Result<Task<FantomasResponse>, FantomasServiceError>) =
     match result with
     | Ok t -> t
     | Error FantomasServiceError.FileDoesNotExist -> fileNotFoundResponse filePath
     | Error FantomasServiceError.FilePathIsNotAbsolute -> fileNotAbsoluteResponse filePath
     | Error FantomasServiceError.DaemonNotFound -> daemonNotFoundResponse filePath
+    | Error FantomasServiceError.CancellationWasRequested -> cancellationWasRequestedResponse filePath
 
 type LSPFantomasService() =
     let cts = new CancellationTokenSource()
     let agent = createAgent cts.Token
 
     interface FantomasService with
         member this.Dispose() =
-            agent.Post Reset
-            cts.Cancel()
+            if not cts.IsCancellationRequested then
+                agent.PostAndReply Reset
+                cts.Cancel()
 
         member _.VersionAsync(filePath, ?cancellationToken: CancellationToken) : Task<FantomasResponse> =
-            getFolderFor filePath
+            isCancellationRequested cts.IsCancellationRequested
```

Comment:
in addition to these checks at the start of the methods, I'd also change the `cancellationToken = orDefaultCancellationToken cancellationToken` logic in them to use `cts.Token` as the token if none is provided.

Reply:
Good catch.